src/datahike/index/hitchhiker_tree/upsert.cljc

```diff
@@ -2,54 +2,86 @@
+      (= s (/ (* m (+ 1 m)) 2)))
+    false))
+
+(def prefix? (memoize increase-by-one?))
+
+(defn- max-t
+  "Returns the key with max 't' component."
+  [ks]
+  (when (seq ks)
+    (apply max-key #(nth % 3) ks)))
+
+(defn mask [new indices]
+  (reduce (fn [mask pos]
+            (assoc mask pos (nth new pos)))
+          [nil nil nil nil]
+          indices))
+
 (defn old-key
   "Returns the old version of the given 'new' key if it exists in 'old-keys'.
-  If there are multiple old versions, the one with the biggest transaction time is returned."
-  [old-keys new]
-  (let [[a b _ _] new]
-    (when (seq old-keys)
-      (when-let [candidates (subseq old-keys >= [a b nil nil])]
+  If there are multiple old versions, the one with the biggest transaction time is returned.
+  'indices' is a vector of integer indicating which positions in keys are significant,
+  i.e., [0 2] means that the first and third entry in the key are used for filtering."
+  [old-keys new indices]
+  (when (seq old-keys)
+    (let [mask (mask new indices)]
+      (when-let [candidates (subseq old-keys >= mask)]
         (->> candidates
              (map first)
-             (take-while #(and (= a (first %)) (= b (second %))))
-             reverse
-             first)))))
+             ((if (prefix? indices) take-while filter)
+              #(reduce (fn [_ i]
+                         (if (= (nth % i) (nth new i))
+                           true
+                           (reduced false)))
+                       true
+                       indices))
+             max-t)))))
 
 (defn remove-old
   "Removes old key from the 'kvs' map using 'remove-fn' function if 'new' and 'old' keys' first two entries match."
-  [kvs new remove-fn]
-  (when-let [old (old-key kvs new)]
+  [kvs new remove-fn indices]
+  (when-let [old (old-key kvs new indices)]
     (remove-fn old)))
 
-(defrecord UpsertOp [key ts]
+(defrecord UpsertOp [key ts indices]
   op/IOperation
   (-insertion-ts [_] ts)
   (-affects-key [_] key)
   (-apply-op-to-coll [_ kvs]
-    (-> (or (remove-old kvs key (partial dissoc kvs)) kvs)
+    (-> (or (remove-old kvs key (partial dissoc kvs) indices) kvs)
         (assoc key nil)))
   (-apply-op-to-tree [_ tree]
     (let [children  (cond
                       (tree/data-node? tree) (:children tree)
                       :else (:children (peek (tree/lookup-path tree key))))]
-      (-> (or (remove-old children key (partial tree/delete tree)) tree)
+      (-> (or (remove-old children key (partial tree/delete tree) indices) tree)
           (tree/insert key nil)))))
 
 (defn old-retracted
   "Returns a new datom to insert in the tree to signal the retraction of the old datom."
-  [kvs key]
-  (when-let [old (old-key kvs key)]
+  [kvs key indices]
+  (when-let [old (old-key kvs key indices)]
     (let [[a b c _] old
           [_ _ _ nt] key]
       ;; '-' means it is retracted and 'nt' is the current transaction time.
       [a b c (- nt)])))
```

Comment:
Do we use negative transaction times for retractions in general?

Reply:
That's what I always thought. @kordano Can you confirm this?